APIKit/URLEncodedSerialization.swift

```diff
@@ -9,52 +9,51 @@ private func unescape(string: String) -> String {
-        var dictionary: [String: AnyObject]?
-        
-        if let string = NSString(data: data, encoding: encoding) as? String {
-            dictionary = [String: AnyObject]()
-            
-            for pair in string.componentsSeparatedByString("&") {
-                let contents = pair.componentsSeparatedByString("=")
-                
-                if contents.count == 2 {
-                    dictionary?[contents[0]] = unescape(contents[1])
-                }
-            }
+    public enum Error: ErrorType {
+        case CannotGetStringFromData
+        case CannotGetDataFromString
+        case CannotCastObjectToDictionary
+        case InvalidFormatString
+    }
+
+    public class func objectFromData(data: NSData, encoding: NSStringEncoding) throws -> [String: String] {
+        guard let string = NSString(data: data, encoding: encoding) as? String else {
+            throw Error.CannotGetStringFromData
         }
-        
-        if dictionary == nil {
-            let userInfo = [NSLocalizedDescriptionKey: "failed to decode urlencoded string."]
-            error.memory = NSError(domain: APIKitErrorDomain, code: 0, userInfo: userInfo)
+
+        var dictionary = [String: String]()
+        for pair in string.componentsSeparatedByString("&") {
+            let contents = pair.componentsSeparatedByString("=")
+
+            guard contents.count == 2 else {
+                throw Error.InvalidFormatString
+            }
+
+            dictionary[contents[0]] = unescape(contents[1])
         }
-        
+
         return dictionary
     }
     
-    public class func dataFromObject(object: AnyObject, encoding: NSStringEncoding, error: NSErrorPointer) -> NSData? {
-        let string = stringFromObject(object, encoding: encoding)
-        let data = string.dataUsingEncoding(encoding, allowLossyConversion: false)
-        
-        if data == nil {
-            let userInfo = [NSLocalizedDescriptionKey: "failed to decode urlencoded string."]
-            error.memory = NSError(domain: APIKitErrorDomain, code: 0, userInfo: userInfo)
+    public class func dataFromObject(object: AnyObject, encoding: NSStringEncoding) throws -> NSData {
+        let string = try stringFromObject(object, encoding: encoding)
+        guard let data = string.dataUsingEncoding(encoding, allowLossyConversion: false) else {
+            throw Error.CannotGetDataFromString
         }
-        
+
         return data
     }
     
-    public class func stringFromObject(object: AnyObject, encoding: NSStringEncoding) -> String {
-        var pairs = [String]()
-        
-        if let dictionary = object as? [String: AnyObject] {
-            for (key, value) in dictionary {
-                let string = (value as? String) ?? "\(value)"
-                let pair = "\(key)=\(escape(string))"
-                pairs.append(pair)
-            }
+    public class func stringFromObject(object: AnyObject, encoding: NSStringEncoding) throws -> String {
```

Comment:
How about changing the signature to `public class func stringFromDictionary(dictionary: [String: AnyObject]) -> String`, and then casting from `AnyObject` to dictionary would be handled in `dataFromObject()`.

Removing `encoding` parameter is already pointed out in #51.